### PR TITLE
Stabilize V15 Metadata

### LINKED ISSUE #12939: Metadata V15 breaking change: Features to include in V15
We would like to include the Runtime API info in the metadata and create a V15 release.
The metadata has been stable for a while and would like to minimize the number
of breaking changes from the user's perspective.

Considering this is a breaking change for the metadata, what other features would we
like to include in the metadata V15 before making a release?

## Metadata with Runtime API

To help client-side applications make runtime API calls, we would like to include
the runtime API information in the metadata (for context: https://github.com/paritytech/substrate/issues/11648).

This feature would help tools like [subxt](https://github.com/paritytech/subxt/issues/631), [polkadot-js](https://github.com/polkadot-js/api/issues/4930), and [capi](https://github.com/paritytech/capi/issues/228) provide a uniform API into the runtime.

Furthermore, after merging the [RPC Spec V2 API](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead.html) into Substrate, the current RPC methods
calling into the runtime will be replaced by `chainHead_unstable_call` (the equivalent of the
current `state_call` method). One example is fetching the metadata via `state_getMetadata`
that would be replaced by a runtime API call to `Metadata_metadata` without parameters.

Discovering which runtime function is available, providing the parameters, and interpreting the
result into the proper shape are tedious and prone to error steps. Including the runtime API info
in the metadata would improve the user experience.

The `decl_runtime_apis` from Substrate will be modified to include the runtime API information in the metadata.

// CC: @jsdw @niklasad1 @bkchr @ascjones @athei @harrysolovay

### PR DESCRIPTION
**Pull request by lexnv**
This PR stabilizes the V15 metadata.

### Metadata V15 changes

- https://github.com/paritytech/substrate/pull/13287
  - Expose APIs to easily update the metadata in the future. This includes `metadata_at_version` and `metadata_versions`
  - Unstable metadata is included under version `u32::MAX` until stabilization
  
- https://github.com/paritytech/substrate/pull/13302
  - Collect type information about the runtime API functions.
  - Allows users to generate APIs to call directly into the runtime https://github.com/paritytech/subxt/pull/918

- https://github.com/paritytech/substrate/pull/14143
  - Add overarching `Call`, `Error` and `Event` enums to the metadata to ease decoding. 
  - To achieve, an extra step of expansion is added to the  `construct_runtime!` for it to propagate default pallet parts 
  
- https://github.com/paritytech/substrate/pull/14123
  - Modify the extrinsic metadata to include type information for `Call`, `Address`, `Signature` and `Extra` types
  
- https://github.com/paritytech/frame-metadata/pull/61
  - Adds support for custom types without breaking changes. 
  - https://github.com/paritytech/polkadot-sdk/issues/182 would use these fields to help the deprecation process
  - Could be used for parachains to include specific information

Closes: https://github.com/paritytech/substrate/issues/12939

@paritytech/subxt-team

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3092396

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3092395

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3092394

**Review by bkchr — APPROVED** [review]

File: primitives/metadata-ir/src/lib.rs [hunk 1]
```diff
@@ -65,7 +58,13 @@ pub fn supported_versions() -> sp_std::vec::Vec<u32> {
 
 /// Transform the IR to the latest stable metadata version.
 pub fn into_latest(metadata: MetadataIR) -> RuntimeMetadataPrefixed {
-	let latest: RuntimeMetadataV14 = metadata.into();
+	let latest: frame_metadata::v15::RuntimeMetadataV15 = metadata.into();
+	latest.into()
+}
+
+/// Transform the IR to the latest stable metadata version.
```
Comment (jsdw) [inline primitives/metadata-ir/src/lib.rs:65]: Nit: comment wrong

```suggestion
/// Transform the IR to metadata version 14.
```

**Review by jsdw — APPROVED** [review]

**Comment by lexnv** [comment]
bot rebase

**Comment by lexnv** [comment]
bot rebase

**Pull request merged**